### PR TITLE
A Complete LoginView with Authentication v0.0.1

### PR DESCRIPTION
**Pull request by MohamedSaiko**
Requesting code review and any recommended updates.

Simulator Screen Shot - iPhone 13 Pro - 2024-03-16 at 11 21 26

### CODE REVIEW

**Comment by akabel92** [comment]
Although you requested review from me yesterday, you are still adding new commits. Please, request a review from me when you are sure that you already finished everything.

**Comment by MohamedSaiko** [comment]
> Although you requested review from me yesterday, you are still adding new commits. Please, request a review from me when you are sure that you already finished everything.

I have already finished committing.
I do request a review.

thanks alot

File: iStore/Cart/CartViewModel.swift [hunk 1]
```diff
@@ -19,9 +19,9 @@ final class CartViewModel: ObservableObject {
         self.cartManager = cartManager
     }
     
-    func loadUserCarts() {
+    func loadUserCarts(withUserId userID: Int) {
         
-        let url = userCartURL + "1"
+        let url = userCartURL + "\(userID)"
```
Comment (akabel92) [inline iStore/Cart/CartViewModel.swift:24]: I am seeing sometimes you are adding new line after as a first line in the function and sometimes not. Please follow the same structure for all functions

File: iStore/Constants.swift [hunk 2]
```diff
@@ -21,6 +21,7 @@ let singleProduct = "https://dummyjson.com/products/"
 let limit = "/?limit=10&skip="
 
 let authenticationUrl = "https://dummyjson.com/auth/login"
+let currentUserURL = "https://dummyjson.com/auth/me"
```
Comment (akabel92) [inline iStore/Constants.swift]: Why do you name it `currentUserURL`? it should be `userURL`. WDYT?

File: iStore/Login/AuthenticationViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    
+    init(authenticationManager: AuthenticationManager, currentUser: CurrentAuthenticatedUser) {
+        self.authenticationManager = authenticationManager
+        self.currentUser = currentUser
+    }
+    
+    func authenticateUser(userName: String, password: String, completion: @escaping () -> Void) {
+        authenticationManager.authenticate(userName: userName, password: password) { [weak self] result in
+            
+            guard let self = self else {
```
Comment (akabel92) [inline iStore/Login/AuthenticationViewModel.swift:25]: You can make it `guard let self` directly. Swift update :D

File: iStore/Login/LoginButton.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+//
+//  Created by Mohamed Sayed on 13.03.24.
+//
+
+import SwiftUI
+
+struct LoginButton: View {
+    
+    @EnvironmentObject var authenticationViewModel: AuthenticationViewModel
+    @EnvironmentObject var navigationCoordinator: NavigationCoordinator
```
Comment (akabel92) [inline iStore/Login/LoginButton.swift]: There is something logically wrong here. Why `LoginButton` should has var for AuthenticationViewModel and var for navigationCoordinator?

Comment (akabel92) [reply iStore/Login/LoginButton.swift]: This is just a button!

File: iStore/Login/LoginView.swift [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                UserNameStack(userName: $userName)
+                
+                PasswordStack(password: $password)
+                
+                if authenticationViewModel.showError {
+                    Text("Please, Enter a valid UserName and Password!")
+                        .foregroundColor(.red)
+                }
+                
+                if authenticationViewModel.showProgress {
```
Comment (akabel92) [inline iStore/Login/LoginView.swift]: This is should be `else if`, right? authenticationViewModel can't show error and progress in the same time

File: iStore/Login/PasswordStack.swift [hunk 6]
```diff
@@ -0,0 +1,24 @@
+//
+//  PasswordStack.swift
+//  iStore
+//
+//  Created by Mohamed Sayed on 13.03.24.
+//
+
+import SwiftUI
+
+struct PasswordStack: View {
```
Comment (akabel92) [inline iStore/Login/PasswordStack.swift]: What do you mean for `PasswordStack`?

File: iStore/Login/RegisterButton.swift [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+//  iStore
+//
+//  Created by Mohamed Sayed on 13.03.24.
+//
+
+import SwiftUI
+
+struct RegisterButton: View {
+    
+    @EnvironmentObject var navigationCoordinator: NavigationCoordinator
```
Comment (akabel92) [inline iStore/Login/RegisterButton.swift]: Also here there is something logically wrong here

File: iStore/Login/RegisterButton.swift [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    
+    var body: some View {
+        Button("Register") {
+            navigationCoordinator.switchView = .registerView
```
Comment (akabel92) [inline iStore/Login/RegisterButton.swift]: This means that the button always will do this line. Why did you make a separate view the button?

File: iStore/Login/UserNameStack.swift [hunk 9]
```diff
@@ -0,0 +1,25 @@
+//
+//  UserNameStack.swift
+//  iStore
+//
+//  Created by Mohamed Sayed on 13.03.24.
+//
+
+import SwiftUI
+
+struct UserNameStack: View {
```
Comment (akabel92) [inline iStore/Login/UserNameStack.swift]: What do you mean for UserNameStack?

File: iStore/Constants.swift (hunk 2 above)
Comment (MohamedSaiko) [reply iStore/Constants.swift]: I think both are ok.

File: iStore/Login/AuthenticationViewModel.swift (hunk 3 above)
Comment (MohamedSaiko) [reply iStore/Login/AuthenticationViewModel.swift:25]: That requires an updated version of xCode.

File: iStore/Constants.swift (hunk 2 above)
Comment (akabel92) [reply iStore/Constants.swift]: current user means for me that you have old user and you will have new user

File: iStore/Login/LoginView.swift (hunk 5 above)
Comment (MohamedSaiko) [reply iStore/Login/LoginView.swift]: I think both are valid, but else if is better.
It wouldn't show both at the same time, as I have already handled Its logic from the AuthenticathionViewModel.

File: iStore/Login/PasswordStack.swift (hunk 6 above)
Comment (MohamedSaiko) [reply iStore/Login/PasswordStack.swift]: It should have been named PasswordView

File: iStore/Login/RegisterButton.swift (hunk 8 above)
Comment (MohamedSaiko) [reply iStore/Login/RegisterButton.swift]: I wanted it to always navigate to the registerView but the backend doesn't support user registration so, I replaced it with button ("Continue as aGuest instead").

The Guest user won't be able to add to cart and their cartView will always be empty.

Simulator Screen Shot - iPhone 13 Pro - 2024-03-18 at 14 02 33

File: iStore/Login/UserNameStack.swift (hunk 9 above)
Comment (MohamedSaiko) [reply iStore/Login/UserNameStack.swift]: It should have been UserNameView

File: iStore/Login/LoginButton.swift (hunk 4 above)
Comment (MohamedSaiko) [reply iStore/Login/LoginButton.swift]: I resolved this but I need another review.

File: iStore/Login/LoginView.swift (hunk 5 above)
Comment (akabel92) [reply iStore/Login/LoginView.swift]: Its related to the compiler process. Two if conditions means that the compiler needs to check both statements. If else not required to check both statements for all cases

**Review by akabel92 — APPROVED** [review]

**Pull request merged**